File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
class Solution {
    public int maxPathSum(TreeNode root) {
        int maxval[]=new int[1];
        maxval[0]=Integer.MIN_VALUE;
        maxpathdown(root,maxval);
        return maxval[0];
    }
    private int maxpathdown(TreeNode node,int maxval[]) {
        if(node==null) return 0;
        int left=Math.max(0,maxpathdown(node.left,maxval));
        int right=Math.max(0,maxpathdown(node.right,maxval));
        maxval[0]=Math.max(maxval[0],left+right+node.val);
        return Math.max(left,right)+node.val;
    }
}

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
class Solution {
public:
    int bfs(TreeNode* root){
        if(!root) return 0;
        int l=max(0,bfs(root->left));
        int r=max(0,bfs(root->right));
        return max(l,r)+root->val;
    }
    int maxPathSum(TreeNode* root) {
        int ans=INT_MIN;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            TreeNode* node=q.front();
            q.pop();
            int l=max(bfs(node->left),0);
            int r=max(bfs(node->right),0);
            ans=max(ans,l+r+node->val);
            
            if(node->left) q.push(node->left);
            if(node->right) q.push(node->right);
        }
        return ans;
    }
};
//dfs
// class Solution {
// public:
//     int maxPathSum(TreeNode* root) {
//         int maxsum=INT_MIN;
//         maxPathDown(root,maxsum);
//         return maxsum;
//     }
//     int maxPathDown(TreeNode* node,int &maxsum){
//         if(node==NULL) return 0;
        
//         int l=max(0,maxPathDown(node->left,maxsum));
//         int r=max(0,maxPathDown(node->right,maxsum));

//         maxsum=max(maxsum,l+r+node->val);
        
//         return max(l,r)+node->val;
//     }
// };